### PR TITLE
cluster: write a frozen kubeconfig when we connect

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas, @lizzthabet,

Please review the following commits I made in branch nicks/minify2:

01f0473accbe5a95cc3514db257935a7ba20b16f (2022-05-06 15:20:37 -0400)
cluster: write a frozen kubeconfig when we connect

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
a lot of this PR is just getting the right API objects to the right places

File: internal/controllers/core/cluster/reconciler.go [hunk 1]
```diff
@@ -357,6 +377,50 @@ func (r *Reconciler) populateK8sMetadata(ctx context.Context, clusterNN types.Na
...
+	}
+
+	obj, err := latest.Scheme.ConvertToVersion(config, latest.ExternalVersion)
+	if err != nil {
+		logger.Get(ctx).Warnf("Converting Kubernetes config: %v", err)
+		return ""
+	}
+
+	printer := printers.YAMLPrinter{}
+	path, err := r.base.ConfigFile(filepath.Join("cluster", nn.Name))
```
Comment (milas) [inline internal/controllers/core/cluster/reconciler.go:401]: Nit: `RuntimeFile` seems more appropriate for this

File: internal/k8s/env.go [hunk 2]
```diff
@@ -12,33 +10,52 @@ import (
...
+		context, ok := config.Contexts[config.CurrentContext]
+		if ok {
+			context.Namespace = string(namespaceOverride)
 		}
 	}
 
-	return &config, nil
+	err = clientcmd.ConfirmUsable(config, config.CurrentContext)
+	if err != nil {
+		return APIConfigOrError{Error: errors.Wrap(err, "Overriding Kubernetes context")}
```
Comment (milas) [inline internal/k8s/env.go:48]: Minor: this is always happening now, so a more generic message like "Loading Kubernetes config" might be more applicable

**Review by milas — APPROVED** [review]
Woooo

File: internal/controllers/core/cluster/reconciler.go (hunk 1 above)
Comment (nicks) [reply internal/controllers/core/cluster/reconciler.go]: done

File: internal/k8s/env.go (hunk 2 above)
Comment (nicks) [reply internal/k8s/env.go]: done

**Pull request merged**